src/libsyntax/ext/base.rs

```diff
@@ -146,6 +156,173 @@ impl Into<Vec<Annotatable>> for Annotatable {
     }
 }
 
+pub trait ProcMacro {
+    fn expand<'cx>(&self,
+                   ecx: &'cx mut ExtCtxt,
+                   span: Span,
+                   ts: TokenStream)
+                   -> TokenStream;
+}
+
+impl<F> ProcMacro for F
+    where F: Fn(TokenStream) -> TokenStream
+{
+    fn expand<'cx>(&self,
+                   _ecx: &'cx mut ExtCtxt,
+                   _span: Span,
+                   ts: TokenStream)
+                   -> TokenStream {
+        // FIXME setup implicit context in TLS before calling self.
+        (*self)(ts)
+    }
+}
+
+pub trait AttrProcMacro {
+    fn expand<'cx>(&self,
+                   ecx: &'cx mut ExtCtxt,
+                   span: Span,
+                   annotation: TokenStream,
+                   annotated: TokenStream)
+                   -> TokenStream;
+}
+
+impl<F> AttrProcMacro for F
+    where F: Fn(TokenStream, TokenStream) -> TokenStream
+{
+    fn expand<'cx>(&self,
+                   _ecx: &'cx mut ExtCtxt,
+                   _span: Span,
+                   annotation: TokenStream,
+                   annotated: TokenStream)
+                   -> TokenStream {
+        // FIXME setup implicit context in TLS before calling self.
+        (*self)(annotation, annotated)
+    }
+}
+
+pub struct TokResult<'a> {
+    pub parser: Parser<'a>,
+    pub span: Span,
+}
+
+impl<'a> MacResult for TokResult<'a> {
+    fn make_items(mut self: Box<Self>) -> Option<SmallVector<P<ast::Item>>> {
+        if self.parser.sess.span_diagnostic.has_errors() {
+            return None;
+        }
+
+        let mut items = SmallVector::zero();
+        loop {
+            match self.parser.parse_item() {
+                Ok(Some(item)) => {
+                    // FIXME better span info.
+                    let mut changer = ChangeSpan { span: self.span };
+                    let new_items = changer.fold_item(item);
+                    items.push_all(new_items);
+                }
+                Ok(None) => {
+                    return Some(items);
```

Comment:
`parse_item()` will return `None` if the next token does not begin an item. For example, if a procedural macro returned an identifier token in an item position, it would silently expand into nothing here.

`ext::tt::macro_rules::ParserAnyMacro` solves this problem with `ensure_complete_parse`.
In general, I think it would be a good idea to follow the semantics of `ParserAnyMacro` more closely or even merge `ParserAnyMacro` and `TokResult` (moving the `ChangeSpan` fold into `expand.rs`), at least until we clean up the semantics of the `parser.parse_*` methods used here.